### PR TITLE
Primary model-function-trainer skeleton

### PR DESCRIPTION
**Pull request by nelson-lojo**
This is meant to be a starting point from which to rebuild the functionality of Garg et. al and other desired features during research.

### CODE REVIEW

File: src/models/nearest_neighbors.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import torch
+from core import ContextModel
+
+
+class KNNModel(ContextModel):
+    def __init__(self, n_neighbors, weights="uniform", **kwargs):
+        # should we be picking k optimally
```
Comment (nelson-lojo) [inline src/models/nearest_neighbors.py:7]: not sure what this comment is for tbh

File: src/models/linear.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        super(LassoModel, self).__init__()
+
+        # the l1 regularizer gets multiplied by alpha.
+        self._alpha = alpha
+        self._max_iter = max_iter
+        self.name = f"lasso_alpha={alpha}_max_iter={max_iter}"
+        self.context_length = -1
+
+    # inds is a list containing indices where we want the prediction.
+    # prediction made at all indices by default.
```
Comment (nelson-lojo) [inline src/models/linear.py:74]: out-of-sync comment, remove

File: src/models/gradient_mlp.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                raise ValueError("inds contain indices where xs and ys are not defined")
+
+        preds = []  # predict one for first point
+
+        # i: loop over num_points
+        for i in tqdm(inds):
+            pred = torch.zeros_like(ys[:, 0])
+            model = self._get_new_model()
+            optim = self._opt(model.parameters())
+            # model.to(DEVICE)
```
Comment (nelson-lojo) [inline src/models/gradient_mlp.py:132]: this should be addressed: how should we determine what device to execute on?

File: src/models/gradient_mlp.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            curried_throw(ValueError(f"GDModel does not support \"{loss_name}\" loss function!"))
+        )()
+        
+        self._batch_size = batch_size
+        self._num_steps = num_steps
+
+        self.name = f"gdmodel_model={model_class_name}_model_kwargs={model_class_args}_opt={opt_alg_name}_lr={lr}_bsize={batch_size}_nsteps={num_steps}_loss={loss_name}"
+        self.context_length = -1
+
+    def forward(self, xs, ys, inds=None, verbose=False, print_step=100):
```
Comment (nelson-lojo) [inline src/models/gradient_mlp.py:109]: inds should not exist as an arg here, right?

File: src/models/decision_tree.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+class DecisionTreeModel(ContextModel):
+    def __init__(self, max_depth=None, **kwargs):
+        super(DecisionTreeModel, self).__init__()
+
+        self._max_depth = max_depth
+        self.name = f"decision_tree_max_depth={max_depth}"
+        self.context_length = -1
+
+    # inds is a list containing indices where we want the prediction.
+    # prediction made at all indices by default.
```
Comment (nelson-lojo) [inline src/models/decision_tree.py:14]: out-of-sync comment, should be removed

File: src/models/__init__.py
Comment (nelson-lojo) [inline src/models/__init__.py:1]: check export syntax (again)

File: src/function_classes/mlp.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import torch
+import torch.distributions as D
+
+from torch.distributions import Distribution
```
Comment (nelson-lojo) [inline src/function_classes/mlp.py:4]: unnecessary import

File: src/function_classes/mlp.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        param_event_shape = torch.Size([self.x_dim + self.y_dim,self._hidden_dim])
+
+        param_dist_shape = torch.Size(batch_shape + param_event_shape)
+
+        param_dist = D.Normal( torch.zeros(param_dist_shape), 
+                               torch.ones(param_dist_shape)   )
+
+        return param_dist
+
+    def evaluate(self, x_batch: torch.Tensor, raw_params: torch.Tensor) -> torch.Tensor:
```
Comment (nelson-lojo) [inline src/function_classes/mlp.py:24]: should there be a mechanism for evaluating on cuda/other devices?

File: src/function_classes/linear.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
+import torch
+import torch.distributions as D
+
+from torch.distributions.distribution import Distribution
```
Comment (nelson-lojo) [inline src/function_classes/linear.py:4]: unnecessary import

File: src/function_classes/linear.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+        param_dist_shape = torch.Size(batch_shape + param_event_shape)
+
+        param_dist = D.Normal( torch.zeros(param_dist_shape), 
+                               torch.ones(param_dist_shape)   )
+
+        return param_dist
+
+    def evaluate(self, x_batch: torch.Tensor, params: torch.Tensor) -> torch.Tensor:
+        # TODO: changed this line so it would run, but it's probably wrong!
```
Comment (nelson-lojo) [inline src/function_classes/linear.py:23]: follow-up needed here

File: src/function_classes/decision_tree.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
+
+import torch
+import torch.distributions as dist
+
+from torch.distributions.distribution import Distribution
```
Comment (nelson-lojo) [inline src/function_classes/decision_tree.py:5]: unnecessary import

File: .gitignore [hunk 11]
```diff
@@ -0,0 +1,4 @@
+__pycache__
+.venv
+.vscode
+src/wandb
```
Comment (nelson-lojo) [inline .gitignore:4]: fix missing newline, add `.idea/`, `*.pkl`, and `*.pth`

File: src/train/train.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+            entity=args.wandb.entity,
+            config=args.__dict__,
+            notes=args.wandb.notes,
+            name=args.wandb.name,
+            resume=True,
+        )
+
+    model = build_model(args.model)
+    model.cuda()
+    model.train()
```
Comment (nelson-lojo) [inline src/train/train.py:130]: this should be done in the `<ContextTrainer>.train(...)` method and `model.eval()` should be called just before returning the model @riensou

File: src/train/train.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        if run_id is None:
+            run_id = str(uuid.uuid4())
+
+        out_dir = os.path.join(args.out_dir, run_id)
+        if not os.path.exists(out_dir):
+            os.makedirs(out_dir)
+        args.out_dir = out_dir
+
+        with open(os.path.join(out_dir, "config.yaml"), "w") as yaml_file:
+            yaml.dump(args.__dict__, yaml_file, default_flow_style=False)
```
Comment (nelson-lojo) [inline src/train/train.py:155]: This is very appealing as an approach to overriding parameters from CLI, although we should decide if that's even a good idea

File: src/train/train.py
Comment (nelson-lojo) [inline src/train/train.py:1]: Should comb through this file in detail to itemize feature parity with this.

At first blush, I see:
- [ ] some cuda stuff? feels like this is to do with Max's comment in class about putting some models on cuda and others not
- [ ] checkpointing
- [ ] test runs (i.e. validate functionality without a bajillion training steps)
- [ ] overriding properties otherwise referenced in the config file

**Review by nelson-lojo — COMMENTED** [review]
A lot to work on

**Comment by riensou** [comment]
functionality for training without a curriculum seems to be broken in parse.py at the moment (i.e. setting x_dim and seq_len to a number in the yml file instead of a !curriculum)

File: src/parse.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+        seq_len = stage['seq_len']
+        x_curriculum_dim = stage['x_dim']
+
+        stage['x_dist'] = get_x_distribution(
+            b_size, seq_len, x_dim, stage.get('x_dist', {})
+        )
+
+        stage['model'] = get_model(
+            stage['model'] | { "x_dim" : x_dim }
+        )
```
Comment (riensou) [inline src/parse.py:237]: I could be wrong, but this seems to create a new model for each stage since `get_model` creates a new model (of the same type, but not shared parameters) each time its called.

potential solution:
- add: `model = None` before the for loops
- add after the current line 235: `model = model or stage['model']`
- add after adding the previous line: `stage['model'] = model`

File: src/train/train.py (hunk 12 above)
Comment (riensou) [reply src/train/train.py:130]: right, I think all of train.py is old and just garbage from the Garg et al codebase at this point

File: src/core/context_model.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        try:
+            y_dim = ys.shape[2]
+        except IndexError as e:
+            y_dim = 1
+
+        xy_seq = torch.cat(
+            (xs, 
+             ys.view(ys.shape + (1,))), 
+            dim=2
+        )
```
[the remark this answers is not in the document]
Comment (riensou) [reply src/core/context_model.py:51]: yeah, when I was working with 3d tensors for y last semester, we had to change how interleaves works

File: src/parse.py (hunk 14 above)
Comment (nelson-lojo) [reply src/parse.py:237]: Nice catch! I think making one model outside the loop and referencing it for each stage would be the same:
```python
model = get_model(stages[0]['model'] | { "x_dim" : x_dim })
for stage in stages:
    ...
    stage['model'] = model
    ...
```

The `or` in the first solution proposed might be a bit verbose, since a curriculum necessarily must have a linear ancestry of models before it (so is always equivalent to just one model shared for all stages)

**Comment by nelson-lojo** [comment]
Requires: #3

File: src/train/train.py (hunk 13 above)
Comment (nelson-lojo) [reply src/train/train.py:155]: We aren't going to pursue this, since it might harm reproducibility if we don't fully merge the YAML config with CLI params.

File: src/train/train.py
Comment (nelson-lojo) [reply src/train/train.py:1]: We decided against "test runs" by commiting to front-load all type-checking/initialization/etc with the exception of gradient/eval steps. This should have the same effect as a test run -- checking for simple errors in configs or code -- without introducing a potential branch in code paths between "testing" and actual training/deployment/evaluation

File: src/function_classes/mlp.py (hunk 7 above)
Comment (nelson-lojo) [reply src/function_classes/mlp.py:23]: See #3

File: src/models/gradient_mlp.py (hunk 3 above)
Comment (nelson-lojo) [reply src/models/gradient_mlp.py:132]: See #3

File: environment.yaml
[the remark this answers is not in the document]
Comment (nelson-lojo) [reply environment.yaml:1]: Conda is a must -- turns out the research cluster uses singularity containers which have good infra with conda. We'll keep conda and build containers that come prepped for it

File: src/function_classes/__init__.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+    LinearClassification, 
+    SparseLinearRegression, 
+    QuadraticRegression
+)
+from .mlp import MLPRegression
+from .wrappers import (
+    ModifiedFunctionClass,
+    NoisyRegression,
+    ScaledRegression
+)
```
[the remark this answers is not in the document]
Comment (nelson-lojo) [reply src/function_classes/__init__.py:13]: See #5

**Review by nelson-lojo — COMMENTED** [review]
Not sure if https://github.com/nelson-lojo/in-context/pull/1/files#diff-c4db8ec7e68cf4494985cd75e9b70065f510030548f62a807d7c14c182725660R21 is right, but otherwise lgtm!

File: src/core/context_model.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (nelson-lojo) [reply src/core/context_model.py:51]: See #7

File: src/train/context_trainer.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+                # wandb.log(
+                #     data=log_dict,
+                #     step=i,
+                # )
+                wandb.log(
+                    data=log_dict
+                )
+
+            # TODO: stretch goal: log functions / precompute dataset?
+            # TODO: log the xs, ys?
```
[the remark this answers is not in the document]
Comment (nelson-lojo) [reply src/train/context_trainer.py]: See #12
